### DISCUSSION TITLE
Passthrough host USB device?

### DISCUSSION THREAD

**Discussion by fengye87** [cloud-hypervisor/cloud-hypervisor#3840] [General]
So I was trying to passthrough a host USB device to a cloud hypervisor VM, but when starting the VM, the API call failed with `start instance: request failed: 500 Internal Server Error: VmBoot(VmBoot(DeviceManager(VfioCreate(InvalidPath))))`.

Here's the device config I'm using:

```json
{
   ...
   "devices":[
      {
         "path":"/dev/bus/usb/002/003",
         "iommu":false,
         "id":"usb",
         "pci_segment":0
      }
   ]
}
```

I also tried to replace the path with `/sys/bus/usb/devices/2-2`, got the same error too. Does cloud hypervisor (v21.0) support USB passthrough? If so, what path should I use here?

**Comment by sboeuf** [comment]
Cloud Hypervisor doesn't support USB passthrough.
You must be passing a PCI device that you have previously bound to the `vfio-pci` driver.

**Comment by fengye87** [comment]
@sboeuf Got it. Is it (USB passthrough) on the road map?

**Comment by sboeuf** [comment]
No it's not.

**Comment by blitz** [comment]
Would USB support be interesting for the project at all? Especially in edge scenarios USB passthrough is very useful and we might have some time to work on this.

**Reply by liuw** [reply]
What do you have in mind? Do you want to passthrough the USB controller itself or emulate one?

**Reply by blitz** [reply]
Passthrough of a PCI XHCI controller should already work. I'm interested in passing through individual USB devices. This would necessitate emulating a USB controller, yes.

**Reply by alyssais** [reply]
Worth noting that [it seems virtio-usb is in development](https://lore.kernel.org/virtio-comment/20240523094134.554857-1-aiswarya.cyriac@opensynergy.com/).  That would probably be a lot more appealing for Cloud Hypervisor than emulating a USB controller.

**Reply by blitz** [reply]
@alyssais Nice!

@rbradford also just told me about [vfio-user](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/vfio-user.md), which would also be a good route for implementing USB emulation (and other extensions that might not fit Cloud Hypervisor's mission).

**Reply by alyssais** [reply]
My preferred eventual solution here would be the [standalone vhost-user devices](https://lore.kernel.org/qemu-devel/20230704123600.1808604-1-alex.bennee@linaro.org/) idea that @stsquad suggested a while ago.  It would allow, with very few modifications necessary, a VMM to query all the information it needs about the device from the backend, avoiding the need for the VMM to enumerate every type of vhost-user device it supports.

(This is a lot nicer than a vfio-user device, because vhost-user devices are simpler to implement (just virtio, not full PCI), and because there are lots of vhost-user devices already out there that could be very easily made to support this.)

**Reply by blitz** [reply]
That's true and this proposal would simplify the vhost-user flow, but it's ultimately limited to virtio devices. That's ok for many usecases, but especially for Windows guests it's really convenient to be able to emulate standard devices that Windows already knows how to drive. So from my point, both the vhost-user standalone extensions and vfio-user are super useful!

**Reply by alexellis** [reply]
We would be interested in supporting USB devices for SlicerVM.com users - mostly home users who have some kind of USB device like a TPU or radio for Flightaware.

I hadn't thought of passing a whole USB controller through VFIO.. that's an interesting idea if supported well in a computer's IOMMU groups

**Reply by alyssais** [reply]
You might also be interested in https://github.com/cyberus-technology/usbvfiod.

**Reply by alexellis** [reply]
Thanks that does look worth a try
